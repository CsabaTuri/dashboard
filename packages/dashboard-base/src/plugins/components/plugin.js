import { Component } from "react"

const outletsByName = {}
const pluginByOutlet = {}

export class Outlet extends Component {
  constructor(props) {
    super(props)
    if (!props.name) throw new Error("Outlet must have a name")
  }

  componentWillMount() {
    outletsByName[this.props.name] = this
  }

  componentWillUnmount() {
    delete outletsByName[this.props.name]
  }

  render() {
    const plugin = pluginByOutlet[this.props.name]
    return plugin ? plugin.content() : this.props.children || null
  }
}

export class Plugin extends Component {
  constructor(props) {
    super(props)
    if (!props.outlet) throw new Error("Plugin must have an outlet")
  }

  componentWillMount() {
    pluginByOutlet[this.props.outlet] = this
  }

  componentWillUnmount() {
    delete pluginByOutlet[this.props.outlet]
  }

  content() {
    return this.props.children
  }

  render() {
    const outlet = outletsByName[this.props.outlet]
    if (outlet) outlet.forceUpdate()
    return null
  }
}
