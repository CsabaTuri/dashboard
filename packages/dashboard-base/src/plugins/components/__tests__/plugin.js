import React from "react"
import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"

import { Outlet, Plugin } from "../plugin"

it("should render a plugin", () => {
  const comp = shallow(
    <Outlet name="plugin-goes-here">
      <h1>Default content</h1>
    </Outlet>
  )
  expect(shallowToJson(comp)).toMatchSnapshot()

  const plugin = shallow(
    <Plugin name="fake-plugin" outlet="plugin-goes-here">
      <h1>I will appear at "plugin-goes-here"</h1>
    </Plugin>
  )

  comp.update() // Force test wrapper to fetch plugin changes on the outlet component
  expect(shallowToJson(plugin)).toMatchSnapshot()
  expect(shallowToJson(comp)).toMatchSnapshot()
})

it("should render an empty outlet if no default content", () => {
  const comp = shallow(<Outlet name="plugin-goes-here" />)
  expect(shallowToJson(comp)).toMatchSnapshot()
})
