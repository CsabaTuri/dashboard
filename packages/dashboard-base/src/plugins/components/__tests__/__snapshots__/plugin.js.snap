// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a plugin 1`] = `
<h1>
  Default content
</h1>
`;

exports[`should render a plugin 2`] = `null`;

exports[`should render a plugin 3`] = `
<h1>
  Default content
</h1>
`;

exports[`should render an empty outlet if no default content 1`] = `null`;
